gen_proc_macro/src/common.rs

```diff
@@ -0,0 +1,107 @@
+                            let ex: Ident =
+                                syn::parse2(yd_expr.mac.tokens.clone()).expect("");
+                            quote! { #ex }
+                        }
+                        _ => panic!("bug found index of yield in stmts but no yield"),
+                    };
+                    let co_call = quote! { co.yield_(#ident).await; };
+                    let cc: Stmt = parse2(co_call).expect("parse of Stmt failed");
+                    loop_stmts.remove(idx);
+                    loop_stmts.insert(idx, cc);
+                }
+            }
+            Stmt::Expr(Expr::While(loopy)) => {
+                let loop_stmts = &mut loopy.body.stmts;
+                let yield_idx = loop_stmts.iter().position(|loop_stmt| {
+                    match loop_stmt {
+                        Stmt::Item(Item::Macro(m)) => {
+                            if let Some(id) = m.mac.path.get_ident() {
+                                println!("{:?}", m);
+                                id == "yield_"
+                            } else {
+                                false
+                            }
+                        }
+                        _ => false,
+                    }
+                });
+                if let Some(idx) = yield_idx {
+                    let ident = match loop_stmts.get(idx) {
+                        Some(Stmt::Item(Item::Macro(yd_expr))) => {
+                            let ex: Ident =
+                                syn::parse2(yd_expr.mac.tokens.clone()).expect("");
+                            quote! { #ex }
+                        }
+                        _ => panic!("bug found index of yield in stmts but no yield"),
+                    };
+                    let co_call = quote! { co.yield_(#ident).await; };
+                    let cc: Stmt = parse2(co_call).expect("parse of Stmt failed");
+                    loop_stmts.remove(idx);
+                    loop_stmts.insert(idx, cc);
+                }
+            }
+            Stmt::Item(Item::Macro(m)) => {
+                if let Some(id) = m.mac.path.get_ident() {
+                    if id == "yield_" {
+                        let ex: Ident = syn::parse2(m.mac.tokens.clone())
+                            .or_else::<(), _>(|e| {
+                                abort!(
+                                    m.span(),
+                                    format!("must yield explicit `()` {}", e)
+                                )
+                            })
+                            .unwrap();
+                        let co_call = quote! { co.yield_(#ex).await; };
+                        let cc: Stmt = parse2(co_call).expect("parse of Stmt failed");
+                        std::mem::replace(stmt, cc);
+                    }
+                } else {
+                    abort!(m.span(), "must yield explicit `()`")
+                }
+            }
+            _ => continue,
```

Comment:
There are a ton of expressions this won't cover.

Here are two examples (modifying one of the tests you added, the only thing I changed was the `yield_` statements and asserts):

<details>
<summary>Details</summary>

```rust
#[test]
fn sync_proc_macro_fn() {
    #[yielder_fn_sync(u8)]
    async fn odds() {
        for n in (1_u8..).step_by(2).take_while(|&n| n < 10) {
            yield_! { n };
            yield_! { n };
        }
    }
    let gen = Gen::new(odds);
    let res = gen.into_iter().collect::<Vec<_>>();
    assert_eq!(vec![1, 1, 3, 3, 5, 5, 7, 7, 9, 9], res)
}
```

</details>

<details>
<summary>Details</summary>

```rust
#[test]
fn sync_proc_macro_fn() {
    #[yielder_fn_sync(u8)]
    async fn odds() {
        for n in (1_u8..).step_by(2).take_while(|&n| n < 10) {
            if true {
                yield_! { n };
            }
        }
    }
    let gen = Gen::new(odds);
    let res = gen.into_iter().collect::<Vec<_>>();
    assert_eq!(vec![1, 3, 5, 7, 9], res)
}
```

</details>

Unfortunately I think the current approach will not work. There are dozens of `Expr` types and no way you can/should have to cover them all. Instead you should use the visitor pattern, so traversal is automatically handled for you. I believe [visit_mut](https://docs.rs/syn/1.0.11/syn/visit_mut/index.html) is the right place to start.

Reply:
Yeah complexity was slowly blowing up in my face. I will push what I have fixed and work on this hopefully a day or two will be enough.